x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -120,6 +129,10 @@ synchronized void updateWithFailure(Exception exc) {
         failIfFrozen();
         // copy the response headers from the current context
         this.responseHeaders = threadContext.getResponseHeaders();
+        //We may have already received some partial results, in which case the number of successful shards reflects that despite the search
+        //has failed entirely at a later stage. We should consider all shards as failed given that none of them was able to e.g. fetch
+        //skipped shards are considered successful though
+        this.successfulShards = this.skippedShards;
```

Comment:
I believe this makes what gets returned more consistent with what _search does, though we lose information on how many shards the partial results come from. We possibly need to expand the info that we return if we want to better represent this scenario where there are partial results yet the whole search has failed hence stopped.

Reply:
Why do we need to update these informations ? They are not used anymore when the response is final. I think https://github.com/elastic/elasticsearch/pull/55683 has the more straightforward approach of keeping the final response as is.

Reply:
I think that this is the reason why you can end up with situations like successful: 3 failed: 3 total: 3 .

Reply:
I would rather want successful: 0 failed: 3 total:3 ( I left skipped out for simplicity)  

I have yet to look at the linked PR, will do

Reply:
ok I am now up-to-date, I agree with you on `updateFinalResponse`, no need to touch successful shards anymore, but you commented on updateWithFailure which still has the problem of returning a shards header that search does not return? Hence shall we signal how many shards have returned partial results although all of them later failed, or shall we just zero the successful shards that makes things more consistent with search? Otherwise I think we would need to add more details on which phase has failed to the response to be more accurate... I will update and merge conflicts.

Reply:
I think we'll solve this discrepancy when wee add partial top hits in the response. In the meantime I don't think we should reset the successful shards, it would be weird to have `successful: 0` (assuming no shards were skipped) but some partial results in the response ?

Reply:
I think it's weird either way. I cant make up my mind on which way is less weird :) I will revert this bit then and maybe add a comment that explains what happens today.

Reply:
maybe one thing we could rather do in this case is to reset shard failures, as they have caused a fatal failure and don't need to be returned as part of the search response too, they are already in the outer error section. That way we would have failed: 0 in the inner search response, which makes more sense as it's a snapshot of the results before the failure happened. Not too sure though if this may end up causing problems in other scenarios. Doing nothing is also fine with me, as long as we are aware that the shards section can be weird at times.

Reply:
What about ignoring fetch shard failures when building partial responses ?
We never return partial top hits if the response is partial so the successful and failure counts should only reflect the query phase ? This will change when we add the support for partial top hits but that's the least confusing solution I can think of.